src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,197 @@
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodosContext } from '../Store/Store';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = React.memo(({ todo }) => {
+  const {
+    todos,
+    setTodos,
+    loading,
+    isCompletedAll,
+    setIsCompletedAll,
+    deleteTodo,
+    errorMessage,
+    pressClearAll,
+  } = useContext(TodosContext);
+
+  const [isCompleted, setIsCompleted] = useState(todo.completed);
+  const [isEditing, setIsEditing] = useState(false);
+  const [editTitle, setEditTitle] = useState(todo.title);
+  const todoField = useRef<HTMLInputElement>(null);
+  const [isDeleting, setIsDeleting] = useState(false);
+
+  const handleEditTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setEditTitle(event.target.value);
+  };
+
+  const updateTodo = useCallback((updatedTodo: Todo) => {
+    const updatedTodos = todos.map(upTodo => (
+      upTodo.id === updatedTodo.id
+        ? { ...updatedTodo }
+        : upTodo
+    ));
+
+    setTodos(updatedTodos);
+  }, [todos]);
+
+  const handleCheckbox = () => {
+    setIsCompleted(!isCompleted);
+    const updatedTodo = { ...todo, completed: !isCompleted };
+
+    updateTodo(updatedTodo);
+  };
+
+  const handleDeleteTodo = () => {
+    setIsDeleting(true);
+    deleteTodo(todo.id);
+  };
+
+  useEffect(() => {
+    if (errorMessage) {
+      setIsDeleting(false);
+    }
+  }, [errorMessage]);
+
+  const applyEditing = () => {
+    if (editTitle.length === 0) {
+      handleDeleteTodo();
+
+      return;
+    }
+
+    if (editTitle !== todo.title) {
+      const updatedTodo = { ...todo, title: editTitle };
+
+      updateTodo(updatedTodo);
+    }
+
+    setIsEditing(false);
+  };
+
+  const handleEdit = (event: React.FormEvent) => {
+    event.preventDefault();
+    applyEditing();
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setEditTitle(todo.title);
+      setIsEditing(false);
+    }
+  };
+
+  useEffect(() => {
+    if ((isEditing || errorMessage) && todoField.current) {
+      todoField.current.focus();
+    }
+  }, [isEditing, errorMessage]);
+
+  useEffect(() => {
+    const completedAll = todos.every(completedTodo => completedTodo.completed);
+
+    if (completedAll && todos.length !== 0) {
+      setIsCompletedAll(true);
+    }
+
+    if (!completedAll) {
+      setIsCompletedAll(null);
+    }
+  }, [isCompleted, todos]);
+
+  useEffect(() => {
+    if (isCompletedAll) {
+      setIsCompleted(isCompletedAll);
+
+      const updatedTodos = todos.map(upTodo => (
+        { ...upTodo, completed: isCompletedAll }
+      ));
+
+      setTodos(updatedTodos);
+    }
+
+    if (isCompletedAll === false) {
+      setIsCompleted(isCompletedAll);
+
+      const updatedTodos = todos.map(upTodo => (
+        { ...upTodo, completed: isCompletedAll }
+      ));
+
+      setTodos(updatedTodos);
+    }
+  }, [isCompletedAll]);
```

Comment:
this should be refactored, and it can be done with one useEffect I suppose
please move it to the context